README.md
# npm-travis
[![NPM][NPM1]][NPM2]

[![Build Status][1]][2] [![dependency status][3]][4] [![dev dependency status][5]][6]

This tool allows integrating [Travis] test runs with npm-based
workflows.  The original use case was triggering [Travis] builds
from [Jenkins] in Wikimedia's [Gerrit] code review tool.

The basic idea is that the `npm-travis` binary will push a throwaway
branch to [github] containing the current git HEAD (which is under
review and has not yet been merged). It then uses the [Travis API]
to monitor the build triggered by this push, stream the logs as
they appear (for real-time progress feedback), and then exit with
an error code corresponding to the passed/failed status of the
travis build (after deleting the temporary branch).

In WMF's actual use, we push to [Gerrit]'s repository, which then gets
mirrored to github.  (You can specify the `--remote` option to the
CLI to specify a particular push target.)  This adds a little bit
of latency, but it prevents our Gerrit-to-github synchronization job
from attempting to "sync" and remove our temporary branch while we're
in the middle of using it.  This also allows us to use [Gerrit]'s
access controls to give `npm-travis` the ability to push only to
branches prefixed with `npm-travis/`.

## Suggested use
```
$ cd my-node-tool
$ npm install --save-dev npm-travis
```
Then add the following to the `package.json` of *my-node-tool*:
```
  "scripts": {
    "travis": "npm-travis"
  }
```

Add a job to trigger `npm run travis` to verify a submitted patch (for
example, using [Jenkins Job Builder]/[Zuul]).  The exit status of this
job will mirror the Travis build status, and the console log will
contain the console output of all Travis jobs (build configurations)
associated with this build.

**For non-node.js projects**, you can install the npm-travis tool globally
(`npm install -g npm-travis`) and just invoke the `npm-travis` binary
directly in the Jenkins job (rather than using `npm run`).

## Advanced use

If you would like to have a separate Jenkins job for every Travis
build configuration, use the `--job` option:
```
  "scripts": {
    "travis-1": "npm-travis --job 1",
    "travis-2": "npm-travis --job 2",
    /* etc */
  }
```
If Travis triggers *N* jobs per build (for example, `7.1`, `7.2`, ... `7.N`)
then there should be *N* scripts here.

Jenkins should then trigger `npm run travis-1`...`npm run travis-N`.
Each job's exit status will mirror the Travis passed/failed status,
and the Jenkins console log will contain the Travis log output for
only the specific job/build configuration.

## Configuration

The `npm-travis` tool needs two pieces of information: the name of
the [github] repository which is triggering the Travis build, and
a git remote which it can push to.  The name of the git remote
(referred to as `<remote>` below) is given by the `--remote`
command-line option, defaulting to `github`.  It then attempts to
discover the github repository name as follows:

1. If there is a repository name given on the command line, use that.
2. Otherwise, use `git remote show <remote>` to look at the push URL.
If the push URL looks like a github url, extract the repository name from it.
3. Otherwise, use `git remote show origin` and see if its push URL
looks like a github URL, extracting a repository name if so.
4. Otherwise, look in `package.json` in the current working directory, and see
if its `repository` field looks like a github URL, and extract a
repository name.
5. Fail.

It then checks whether the git remote `<remote>` exists, and if it
does not it will create it using:
```
git remote add <remote> git@github.com:<repository name>
```

In the common case where you've cloned the code from github, this just
works.  For WMF's setup, we want to push to [Gerrit], but the Gerrit
push URL doesn't map in a simple way to the [github] repository name.
So we create a git remote named `gerrit` ahead of time, and then
invoke `npm-travis --remote gerrit <repository name>`, explicitly
giving the github repository name on the command line.

## Security

Note that running `npm-travis` pushes a branch to the github
repository for your project, and therefore that task must be given
write access to the project on github.  Furthermore, if the task is
triggered on patch submission, anyone with permission to submit
patches can create a (short-lived) branch on your repository and run
arbitrary code on a Travis instance.  Be cautious.  You might want to
point `npm-travis` at a fork or mirror of your github repository.

In addition, Travis "secure variables" are normally
[disabled for cross-origin pull requests](http://blog.travis-ci.com/2013-06-10-secure-env-in-pull-requests/);
but since `npm-travis` uses same-origin branches rather than pull
requests, it is not safe to use Travis "secure variables" in
repositories using `npm-travis`.

## License

MIT license; see [LICENSE](./LICENSE).

(c) 2014 by C. Scott Ananian

[github]:  https://github.com
[Travis]:  https://travis-ci.org/
[Travis API]: http://docs.travis-ci.com/api/
[Jenkins]: https://www.mediawiki.org/wiki/Continuous_integration/Jenkins
[Gerrit]:  https://www.mediawiki.org/wiki/Gerrit
[Jenkins Job Builder]: https://www.mediawiki.org/wiki/Continuous_integration/Jenkins_job_builder
[Zuul]:    https://www.mediawiki.org/wiki/Continuous_integration/Zuul

[NPM1]: https://nodei.co/npm/npm-travis.png
[NPM2]: https://nodei.co/npm/npm-travis/

[1]: https://travis-ci.org/cscott/npm-travis.svg
[2]: https://travis-ci.org/cscott/npm-travis
[3]: https://david-dm.org/cscott/npm-travis.svg
[4]: https://david-dm.org/cscott/npm-travis
[5]: https://david-dm.org/cscott/npm-travis/dev-status.svg
[6]: https://david-dm.org/cscott/npm-travis#info=devDependencies


CHANGELOG.md
# npm-travis x.x.x (not yet released)

# npm-travis 1.0.2 (2016-05-05)
* Update npm dependencies; switch to `babybird` Promise implementation.
* Workaround bug in travis' public API to fix streaming. (#1)

# npm-travis 1.0.1 (2015-09-29)
* Switch back to using slashes in branch names, since upstream bug
  https://github.com/travis-ci/travis-api/issues/145 has been fixed.
* Robustness improvements.

# npm-travis 1.0.0 (2014-11-13)
* Initial release.


lib/p.js
// Helpers for promises.
'use strict';
require('core-js/shim');
var Promise = require('./promise.js');

var cp = require('child_process');

var P = module.exports = {};

// My own version of when's nodefn.call with an explicit 'this',
// used for methods.
P.call = function(fn, self) {
	var args = Array.prototype.slice.call(arguments, 2);
	var pfn = Promise.promisify(fn, false, self);
	return pfn.apply(self, args);
};

// Returns a promise for completion after spawning `program`.
P.spawn = function(program, args, options) {
	return new Promise(function(resolve, reject) {
		options = options || {};
		options.childOptions = options.childOptions || {};
		options.childOptions.stdio = options.childOptions.stdio || 'inherit';
		var killTimer = null, killed = false;
		var clearKillTimer = function() {
			if (killTimer) {
				clearTimeout(killTimer);
				killTimer = null;
			}
		};
		var child = cp.spawn(program, args || [], options.childOptions).
			on('exit', function(exitCode, signal) {
				clearKillTimer();
				if (exitCode === 0) {
					return resolve();
				}
				var timeout = killed && /* Maybe we tried, but failed. */
					(signal === 'SIGTERM' || signal === 'SIGKILL');
				var e = new Error(
					program + ' ' + args.join(' ') + ' ' +
					(timeout ? 'exceeded execution time' : 'exited with code ' + exitCode)
				);
				e.code = exitCode;
				e.signal = signal;
				e.timeout = timeout;
				return reject(e);
			}).on('error', function(err) {
				clearKillTimer();
				reject(err);
			});
		if (options.timeout) {
			killTimer = setTimeout(function() {
				killed = true;
				child.kill('SIGTERM');
				killTimer = setTimeout(function() {
					child.kill('SIGKILL');
					killTimer = null;
				}, options.timeout * 2);
			}, options.timeout);
		}
	});
};

// Returns a promise for stdout and stderr
P.execFile = Promise.promisify(
    cp.execFile, [ 'stdout', 'stderr' ], cp
);

// Returns a promise for completion after iterating through the given
// array in parallel.  The function should return a promise for each element.
// This is like map but we throw away the results.
// If the optional `p` parameter is provided, wait for that to resolve
// before starting to process the array contents.
P.forEachPar = function(a, f, p) {
	return Promise.resolve(p).then(function() {
		return a;
	}).then(function(aResolved) {
		return Promise.all(aResolved.map(f));
	});
};

// Returns a promise for completion after iterating through the given
// array in sequence.  The function should return a promise for each element.
// If the optional `p` parameter is provided, wait for that to resolve
// before starting to process the array contents.
P.forEachSeq = function(a, f, p) {
	// The initial value must not be undefined.  Arbitrarily choose `true`.
	p = p ? Promise.resolve(p).return(true) : Promise.resolve(true);
	return Promise.reduce(a, function(curResult, value, index, total) {
		/* jshint unused: vars */
		return f.call(null, value, index, null);
	}, p);
};


lib/retry-request.js
// Wrapper for the 'request' module, which makes it automatically retry
// requests.
'use strict';
require('core-js/shim');
var Promise = require('./promise.js');

var request = require('request');

var DEFAULT_TIMEOUT = 60 * 1000; /* Milliseconds. */
var DEFAULT_RETRIES = 3;
var DEFAULT_DELAY = 0; /* Retry immediately. */
// Limit the number of simultaneous HTTP requests (T73895).
var REQUEST_LIMIT = 0; /* No limit. */

var ReadableStream = require('readable-stream');
var util = require('util');

var cleanOpts = function(opts) {
	// Clone the opts array, and clean up our own properties so that
	// request doesn't get confused (and so request can't scribble on
	// our properties).
	opts = util._extend({}, opts);
	opts.callback = undefined;
	opts.log = undefined;
	opts.retries = undefined;
	opts.delay = undefined;
	opts.stream = undefined;
	return opts;
};

// Doesn't support all of the request.<foo> methods yet, just the main
// function entry point and a modified API for getting a readable stream.
var RetryRequest = module.exports = function(uri, options, callback) {
	// Duplicate the parameter handling of the 'request' module.
	var opts;
	if (typeof uri === 'undefined') {
		throw new Error('undefined is not a valid uri or options object.');
	}
	if ((typeof options === 'function') && !callback) {
		callback = options;
	}
	if (options && typeof options === 'object') {
		opts = util._extend({}, options);
		opts.uri = uri;
	} else if (typeof uri === 'string') {
		opts = {uri: uri};
	} else {
		opts = util._extend({}, uri);
	}
	if (callback) {
		opts.callback = callback;
	}

	/* To quote the request module:
	 * "People use this property instead all the time so why not just
	 * support it."
	 */
	if (opts.url && !opts.uri) {
		opts.uri = opts.url;
		opts.url = undefined;
	}

	// Ok, all the user options are in opts.
	// Add a default timeout and munge the callback slightly.
	if (opts.timeout === undefined) {
		opts.timeout = RetryRequest.DEFAULT_TIMEOUT;
	}
	if (opts.retries === undefined) {
		opts.retries = RetryRequest.DEFAULT_RETRIES;
	}
	if (opts.delay === undefined) {
		opts.delay = RetryRequest.DEFAULT_DELAY;
	}
	// Optionally log retries.
	var log = opts.log || console.error.bind(console);
	var origCb = opts.callback;
	// Promise compatibility.
	var resolve, reject, p = new Promise(function(_resolve, _reject) {
		resolve = _resolve; reject = _reject;
	}).nodify(origCb && function(err, o) {
		// Thunk to traditional node callback API.
		origCb(err, o && (o.response || o.stream), o && (o.body || o.request));
	});

	var req, mkrequest, n = 0;
	var ncallback = function(error, response, body) {
		if (opts.retries > 0 && (error || response.statusCode !== 200)) {
			log('Retrying (' + (++n) + ')', opts.uri, error || response.statusCode);
			opts.retries--;
			opts.timeout *= 2;
			return Promise.delay(opts.delay).then(mkrequest);
		}
		// Promise API.
		if (error) {
			return reject(error);
		}
		return resolve({ response: response, body: body });
	};
	if (!opts.stream) {
		opts.callback = ncallback;
	}
	mkrequest = function() {
		req = request(cleanOpts(opts), opts.callback);
		if (opts.stream) {
			var rstream = req;
			if (/^v0\.8\./.test(process.version)) {
				rstream = new ReadableStream();
				rstream.wrap(req);
			}
			rstream.pause();
			return resolve({ stream: rstream, request: req });
		}
		return p;
	};
	if (RetryRequest.REQUEST_LIMIT) {
		if (RetryRequest._cached_limit !== RetryRequest.REQUEST_LIMIT) {
			RetryRequest._cached_guard = Promise.guard.n(
				(RetryRequest._cached_limit = RetryRequest.REQUEST_LIMIT)
			);
		}
		return Promise.guard(RetryRequest._cached_guard, mkrequest)();
	}
	return mkrequest();
};
RetryRequest.DEFAULT_TIMEOUT = DEFAULT_TIMEOUT;
RetryRequest.DEFAULT_RETRIES = DEFAULT_RETRIES;
RetryRequest.DEFAULT_DELAY = DEFAULT_DELAY;
RetryRequest.REQUEST_LIMIT = REQUEST_LIMIT;


lib/index.js
'use strict';
require('core-js/shim');
var Promise = require('./promise.js');

var program = require('commander');
var json = require('../package.json');

var fs = require('fs');
var Pusher = require('pusher-client');
var url = require('url');

var p = require('./p');
var request = require('./retry-request');

var CHILD_IGNORE = { childOptions: { stdio: 'ignore' } };
var readFile = Promise.promisify(fs.readFile, false, fs);

program
	.version(json.version)
	.usage('[options] [github user/project]')
	.option(
		'--travis-api <url>',
		'Travis API endpoint [https://api.travis-ci.org]',
		'https://api.travis-ci.org'
	).option(
		'--remote <name>',
		'Use the specified git remote for push [github]',
		'github'
	).option(
		'-j, --job <n>',
		'Only follow the specified job (job numbers start at 1)'
	);

program.parse(process.argv);

if (program.args.length > 1) {
	console.error('Too many arguments.');
	return 1;
}

var getUserAgent = function(reponame) {
	return json.name + '/' + json.version + ' (' +
		'node ' + process.version + ' ' + process.platform + ' ' + process.arch +
		')' + (reponame ? ' ' + reponame : '');
};

var gitRemoteRegex =
	/Push\s+URL:\s+(?:https?:\/\/github.com\/|git@github.com:)([^\/]+)\/([^\/]+?)(?:\.git)?\s/;

var getRepoName = function() {
	return Promise.resolve(program.args[0]).then(function(reponame) {
		if (reponame) { return reponame; }
		// Look in .git/config for 'github' remote.
		return p.execFile('git', ['remote','show',program.remote]).then(function(out) {
			var m = gitRemoteRegex.exec(out.stdout);
			if (m) {
				return m[1] + '/' + m[2];
			}
		});
	}).catch(function(e) { /* Ignore. */ }).then(function(reponame) {
		if (reponame) { return reponame; }
		// Look in .git/config for 'origin' remote.
		return p.execFile('git', ['remote','show','origin']).then(function(out) {
			var m = gitRemoteRegex.exec(out.stdout);
			if (m) {
				return m[1] + '/' + m[2];
			}
		});
	}).catch(function(e) { /* Ignore. */ }).then(function(reponame) {
		if (reponame) { return reponame; }
		// Look in ./package.json
		return readFile(
			'./package.json', { encoding: 'utf8' }
		).then(function(data) {
			var json = JSON.parse(data);
			if (json.repository && json.repository.type === 'git') {
				var m =
					/^(https?:\/\/github.com\/)?([^\/]+)\/([^\/]+?)(\.git)?$/.
					exec(json.repository.url || '');
				if (m) {
					return m[2] + '/' + m[3];
				}
			}
		});
	}).catch(function(e) { /* Ignore */ }).then(function(reponame) {
		if (reponame) { return reponame; }
		throw new Error("Can't determine github repo name.");
	});
};

// Ensure that there is a github remote, named 'github'.
var ensureRemote = function(repo) {
	return p.spawn('git', ['remote', 'show', program.remote], CHILD_IGNORE).then(function() {
		/* OK, github already exists. */
	}, function(e) {
		/* Doesn't exist, create it! */
		return p.spawn('git', ['remote','add',program.remote,'git@github.com:' + repo]);
	});
};

// Get the hash of HEAD.
var getGitHead = function() {
	return p.execFile('git', ['rev-parse', 'HEAD']).then(function(out) {
		return out.stdout.trim();
	});
};

// Create the new branch name.
var getBranchName = function(hash) {
	// Use a slash in the branch name so that gerrit can use
	// its access controls to ensure that the npm-travis user can only
	// push to npm-travis/* branches (not branches in general).
	return json.name + '/' + hash.slice(0, 8);
};

// Push to github.
var ensureBranch = function(repo, branchname) {
	return ensureRemote(repo).then(function() {
		return p.spawn('git', ['push', program.remote, 'HEAD:refs/heads/' + branchname], CHILD_IGNORE);
	});
};

var deleteBranch = function(repo, branchname) {
	return ensureRemote(repo).then(function() {
		return p.spawn('git', ['push', program.remote, ':' + branchname], {
			timeout: 60 * 1000,
			childOptions: CHILD_IGNORE.childOptions,
		});
	});
};

var apiRequest = function(path, opts) {
	opts = opts || {};
	if (Array.isArray(path)) {
		path = '/' + path.map(encodeURIComponent).join('/');
	}
	var apiURL = url.resolve(program.travisApi, path);
	var mimeType = 'application/vnd.travis-ci.2+json';
	if (opts.chunked) {
		mimeType += '; chunked=true';
	}
	return request({
		url: apiURL,
		encoding: 'utf8',
		pool: false,
		headers: {
			'User-Agent': getUserAgent(opts.reponame),
			Accept: mimeType,
		},
		retries: opts.retries,
		delay: opts.delay,
		timeout: opts.timeout,
	}).then(function(resp) {
		if (resp.response.statusCode !== 200) {
			throw new Error('Bad status: ' + resp.response.statusCode);
		}
		return opts.raw ? resp.body : JSON.parse(resp.body);
	});
};

var getTravisConfig = function() {
	return apiRequest('/config').get('config');
};

var getTravisRepo = function(reponame) {
	return apiRequest('/repos/' + reponame, { reponame: reponame }).get('repo');
};

var getTravisBuild = function(reponame, repo_id, branch, retry) {
	return apiRequest(['repos', repo_id, 'branches', branch], {
		reponame: reponame,
		timeout: 30 * 1000,
		delay:   5 * 1000, /* Check every 5 seconds... */
		retries: retry ? 3 * 12 : 0, /* ...for up to 3 minutes. */
	}).get('branch');
};

var getTravisJob = function(reponame, job_id) {
	return apiRequest(['jobs', job_id], { reponame: reponame }).get('job');
};

var getTravisLog = function(reponame, log_id, chunked) {
	return apiRequest(['logs', log_id], {
		reponame: reponame,
		chunked: chunked,
	}).get('log');
};

var Log = function() {
	this.chunks = [];
	this.next = 1;
	this._last = null;
};
Log.prototype.has = function(n) {
	return this.chunks[n] !== undefined;
};
Log.prototype.add = function(n, data) {
	this.chunks[n] = data;
	return this.advance();
};
Log.prototype.done = function() {
	return this._last !== null && this.next > this._last;
};
Log.prototype.advance = function() {
	while (this.chunks[this.next] !== undefined) {
		process.stdout.write(this.chunks[this.next++]);
	}
	return this.done();
};
Log.prototype.last = function(n) {
	this._last = n;
	return this.done();
};

var doit = function() {
	var state = {}, pusher;
	var initState = function() {
		return Promise.all([
			getTravisConfig().then(function(config) {
				state.config = config;
				var pusherKey = state.config.pusher.key;
				if (pusherKey === 'key') {
					// Travis' public API is broken!
					// https://github.com/travis-ci/travis-ci/issues/6002
					pusherKey = '5df8ac576dcccf4fd076';
				}
				pusher = new Pusher(pusherKey);
				if (false) { /* Debugging. */
					pusher.connection.on('message', function(message) {
						console.log('GOT PUSHER MESSAGE', message);
					});
				}
			}),
			getGitHead().then(function(hash) {
				state.githash = hash;
				state.branch = getBranchName(hash);
			}),
			getRepoName().then(function(reponame) {
				state.reponame = reponame;
				return getTravisRepo(reponame).then(function(repo) {
					state.repo = repo;
				});
			}),
		]).then(function() {
			// Check to see if builds on this branch already exist.
			return getTravisBuild(state.reponame, state.repo.id, state.branch, false).catch(function(e) {
				// Doesn't yet exist; push to github and then retry query until
				// we see the build show up in travis.
				return ensureBranch(state.reponame, state.branch).then(function() {
					// Clean up this branch on Ctrl-C.
					state.SIGINT = function() {
						console.log('\nCleaning up', state.branch);
						deleteBranch(state.reponame, state.branch).catch(function(e) {
							/* Ignore failure; we're dying here. */
						}).then(function() { process.exit(1); });
					};
					process.once('SIGINT', state.SIGINT);
					// Ok, get the build on this branch.
					return getTravisBuild(state.reponame, state.repo.id, state.branch, true);
				});
			});
		}).then(function(build) {
			// Ok, we've got a build on this branch!
			state.build = build;
			console.log(
				'==> Travis build',
				state.build.number,
				'(',
				'https://' + state.config.host + '/' + state.reponame + '/builds/' + state.build.id,
				')'
			);
			return state;
		});
	};
	var watchJob = function(job_num) {
		return Promise.resolve().then(function() {
			state.job_num = job_num;
			state.job_id = state.build.job_ids[state.job_num - 1];
			console.log(
				'==> Travis job',
				state.build.number + '.' + state.job_num,
				'(',
				'https://' + state.config.host + '/' + state.reponame + '/jobs/' + state.job_id,
				')'
			);
			// Check job #
			if (!state.job_id) {
				throw new Error(
					'Bad job number: ' + state.job_num +
						' (' + state.build.job_ids.length + ' jobs)'
				);
			}
			return getTravisJob(state.reponame, state.job_id);
		}).then(function(job) {
			state.job = job;
			var log = new Log();
			if (/^(passed|failed)$/.test(state.job.state || '')) {
				// Fetch all the logs, easy-peasy.
				return getTravisLog(state.reponame, state.job.log_id, false/*Not chunked*/).then(function(resp_log) {
					log.last(1);
					log.add(1, resp_log.body);
				});
			}
			// Hm, we need to stream the logs.
			var channelName = 'job-' + state.job_id;
			var channel = pusher.subscribe(channelName);
			var streamResolve;
			var handleChunk = function(data) {
				var content = data._log !== undefined ? data._log : data.content;
				if (data.final) {
					// Weird: *Usually* there's one more chunk coming.
					// Wait for it, but no more than 10s.
					log.last(data.number + 1);
					setTimeout(function() {
						if (!log.has(data.number + 1)) {
							log.add(data.number + 1, '\n----\n');
							streamResolve();
						}
					}, 10 * 1000);
				}
				var last = log.add(data.number, content);
				if (last) { streamResolve(); }
			};
			return Promise.all([
				// Fetch chunked log to fill in gaps.
				new Promise(function(resolve, reject) {
					channel.bind('pusher:subscription_succeeded', function() {
						getTravisLog(state.reponame, state.job.log_id, true /* Chunked */).then(function(resp_log) {
							resp_log.parts.forEach(handleChunk);
						}).then(resolve, reject);
					});
				}),
				// Tail log for this job.
				new Promise(function(resolve, reject) {
					streamResolve = resolve;
					channel.bind('job:log', handleChunk);
					channel.bind('job:finished', streamResolve);
				}).then(function() {
					pusher.unsubscribe(channelName);
				}),
			]);
		}).then(function() {
			return getTravisJob(state.reponame, state.job_id);
		}).then(function(job) {
			state.job = job;
			return state.job.state;
		});
	};
	var getBuildStatus = function(retries) {
		return getTravisBuild(
			state.reponame, state.repo.id, state.branch, false
		).then(function(build) {
			if (build.state !== 'started' || retries === 0) {
				return build;
			}
			// There's sometimes a race here: if the build state isn't
			// "passed" or "failed" yet, wait a little bit and try again.
			// In separate jobs mode this isn't unusual, but we still want
			// to be sure that the last one out cleans up the branch.
			console.log('==> ' + json.name + ': ' + build.state + ' (retrying)');
			return Promise.delay(5 * 1000).then(function() {
				return getBuildStatus(retries - 1);
			});
		});
	};
	var cleanup = function() {
		return Promise.resolve().then(function() {
			pusher.disconnect();
			return getBuildStatus(6/*30 seconds*/);
		}).then(function(build) {
			state.build = build;
			// If build is complete (all jobs), we can clean up the branch.
			if (state.SIGINT) {
				process.removeListener('SIGINT', state.SIGINT);
			}
			if (/^(passed|failed)$/.test(state.build.state || '')) {
				return deleteBranch(state.reponame, state.branch).catch(function(e) {
					/* Ignore failure, maybe already deleted. */
				});
			}
		});
	};

	// Ok, put all the pieces together.
	return initState().then(function() {
		if (program.job !== undefined) {
			// Watch one particular job.
			return watchJob((+program.job));
		}
		// Watch all jobs.
		return Promise.reduce(state.build.job_ids, function(_,jid,index) {
			return watchJob(index + 1);
		}, null);
	}).then(function() {
		return cleanup();
	}).then(function() {
		if (program.job !== undefined) {
			// Return our job status.
			return state.job.state;
		}
		// Return our build status.
		return state.build.state;
	});
};

doit().then(function(status) {
	console.log('==> ' + json.name + ': ' + status);
	if (status !== 'passed') {
		process.exit(1);
	} else {
		process.exit(0);
	}
}).done();
